### PR TITLE
Make SHACL regex patterns follow JSON Schema ones

### PR DESCRIPTION
**Pull request by mristin**
We included the regex pattern as-is from the input which caused problems with the regex engines as the patterns in the meta-model are written in a Python dialect (and assuming that the regex engine works on UTF-32 characters). However, most regex engines in the wild operating on SHACL (*e.g.*, Java SHACL validators) use UTF-16 to represent the text and do not support some parts of the Python regex dialect. For example, in the input meta-model specification, we omit the minimum bound 0 (*e.g.*, ``{,4}``), which breaks with the Java regex engine beneath the SHACL validator.

Instead, with this patch, we parse the pattern from the specification and re-render it into the form that we also use in JSON Schema. We pick JSON Schema regex dialect as most SHACL validators in the wild can deal with it, in particular those based on Java as a platform. Hence, we decide to serve this user base with priority.

Discovered in [aas-core-meta issue 342].

[aas-core-meta issue 342]: https://github.com/aas-core-works/aas-core-meta/issues/342

### CODE REVIEW

File: test_data/rdf_shacl/test_main/expected/aas_core_meta.v3/expected_output/shacl-schema.ttl [hunk 1]
```diff
@@ -259,8 +259,8 @@ aas:BlobShape a sh:NodeShape ;
         sh:maxCount 1 ;
         sh:minLength 1 ;
         sh:maxLength 100 ;
-        sh:pattern "^[\\x09\\x0A\\x0D\\x20-\\uD7FF\\uE000-\\uFFFD\\U00010000-\\U0010FFFF]*$" ;
-        sh:pattern "^([!#$%&'*+\\-.^_`|~0-9a-zA-Z])+/([!#$%&'*+\\-.^_`|~0-9a-zA-Z])+([ 	]*;[ 	]*([!#$%&'*+\\-.^_`|~0-9a-zA-Z])+=(([!#$%&'*+\\-.^_`|~0-9a-zA-Z])+|\"(([	 !#-\\[\\]-~]|[\\x80-\\xff])|\\\\([	 !-~]|[\\x80-\\xff]))*\"))*$" ;
+        sh:pattern "^[\\x09\\x0a\\x0d\\x20-\\ud7ff\\ue000-\\ufffd\\U00010000-\\U0010ffff]*$" ;
```
[the remark this answers is not in the document]
Comment (mristin) [reply test_data/rdf_shacl/test_main/expected/aas_core_meta.v3/expected_output/shacl-schema.ttl:262]: Ok, I'll use the exactly same pattern as in Java.

**Comment by mristin** [comment]
@mhrimaz can you please double-check again? Now I made the SHACL schema exactly follow the Java regex dialect.

**Comment by mhrimaz** [comment]
@mristin Thanks for working on this. 
I can't spot any changes. I think the `shacl-schema.ttl` is not updated.

So just to recap, instead of `"^[\\x09\\x0a\\x0d\\x20-\\ud7ff\\ue000-\\ufffd\\U00010000-\\U0010ffff]*$"`
we should have `"^([\\x09\\x0a\\x0d\\x20-\\ud7ff\\ue000-\\ufffd]|\\ud800[\\udc00-\\udfff]|[\\ud801-\\udbfe][\\udc00-\\udfff]|\\udbff[\\udc00-\\udfff])*$"`

**Comment by mristin** [comment]
> @mristin Thanks for working on this. I can't spot any changes. I think the `shacl-schema.ttl` is not updated.
> 
> So just to recap, instead of `"^[\\x09\\x0a\\x0d\\x20-\\ud7ff\\ue000-\\ufffd\\U00010000-\\U0010ffff]*$"` we should have `"^([\\x09\\x0a\\x0d\\x20-\\ud7ff\\ue000-\\ufffd]|\\ud800[\\udc00-\\udfff]|[\\ud801-\\udbfe][\\udc00-\\udfff]|\\udbff[\\udc00-\\udfff])*$"`

I think I forgot to push the latest changes from my local repository. Fixing it tomorrow.

**Comment by mristin** [comment]
@mhrimaz can you please test now? I forgot to add files to the commit. The patterns are now exactly the same as the ones we used in Java.

**Comment by mhrimaz** [comment]
@mristin in java it worked fine. However, it seems that now the pattern is not python friendly.
This is what we have in [JSON Schema](https://github.com/aas-core-works/aas-core-codegen/blob/0f7345e16cf0c847237b3474eaafc935a362d9ca/test_data/jsonschema/test_main/aas_core_meta.v3/expected_output/schema.json#L67):
`"^([\\x09\\x0a\\x0d\\x20-\\ud7ff\\ue000-\\ufffd]|\\ud800[\\udc00-\\udfff]|[\\ud801-\\udbfe][\\udc00-\\udfff]|\\udbff[\\udc00-\\udfff])*$"`
This is what is generated now:
`"^[\\x{09}\\x{0a}\\x{0d}\\x{20}-\\x{d7ff}\\x{e000}-\\x{fffd}\\x{10000}-\\x{10ffff}]*$"`
I am not sure if these two patterns are identical or not, but it doesn't work in python:
image

**Comment by mristin** [comment]
@mhrimaz you can not have a pattern that works both in Python and Java. Java uses `\x{AB}`, `\x{ABCD}` and `\x{ABCD0123}` syntax, Python uses `\xAB`, `\uABCD` and `\UABCD0123`.

If we use the same patterns as in JSON Schema, then they will not work in Python as Python uses UTF-32 to encode strings in memory, but we assumed UTF-16 encoding in JSON schema (`\uABCD\u0123` in JSON Schema equals `\UABCD0123` in Python).

It probably makes most sense to pick Java as SHACL is mostly used by Java folks?

**Comment by mhrimaz** [comment]
We have SHACL validator in various platforms like python ([pySHACL](https://github.com/RDFLib/pySHACL)) and C# ([dotNetRDF](https://dotnetrdf.org/)), ... I can't really say SHACL is only for java, but certainly more popular. 

I am not a regex expert, so if we should choose between java and python in this case, I would go with java. But it would be good to maybe have 2 different files, or atleast have a documentation about this somewhere.

FYI, i tested it with Apache Jena validator and pySHACL.

**Comment by mristin** [comment]
@mhrimaz can you please ask around a bit before we settle for Java?

@sebbader-sap @sebbader what is your take on this?

**Comment by mristin** [comment]
@mhrimaz I generated the patterns now the same as in JSON Schema -- can you please test with the schema from https://github.com/aas-core-works/aas-core-codegen/pull/517/commits/92bcf9ef6301f51f506304f27013f23831238c80?

**Comment by mhrimaz** [comment]
The schema that is in https://github.com/aas-core-works/aas-core-codegen/commit/92bcf9ef6301f51f506304f27013f23831238c80 works in python. However, I am not sure if the regex works as intended and covers every corner cases. 

You can take any valid/invalid example, concatenate it with the ontology, and then pass it to a validator.

If you want to test in python
```
pip3 install pyshacl
pyshacl -s .\shacl-schema.ttl -a -f turtle .\invalid.ttl
```
image


if you want to test in java, first download the [Apache Jena Commands](https://jena.apache.org/download/index.cgi). You can invoke shacl which is within bin/bat folder:
```
<path to jena shacl script in bin or bat folder/shacl> validate -s shacl-schema-1.ttl -d invalid.txt
```
image

For online validation you can use https://shacl-play.sparna.fr/play/validate or https://www.itb.ec.europa.eu/shacl/any/upload


Github doesn't allow .ttl file, change the file format to ttl.
[invalid.txt](https://github.com/user-attachments/files/16762030/invalid.txt)
[maximal.txt](https://github.com/user-attachments/files/16762031/maximal.txt)

Note that the SHACL has also another critical issue related to `DataSpecificationIec61360`. 
[SHACL states](https://github.com/aas-core-works/aas-core-codegen/blob/92bcf9ef6301f51f506304f27013f23831238c80/test_data/rdf_shacl/test_main/expected/aas_core_meta.v3/expected_output/shacl-schema.ttl#L786) that a `rdf:type aas:LangStringDefinitionTypeIec61360 ;` should have the following attribute:
`<https://admin-shell.io/aas/3/0/LangStringDefinitionTypeIec61360/text>`
However in samples it have `<https://admin-shell.io/aas/3/0/AbstractLangString/text>`. But this is another issue anyways.

**Comment by mristin** [comment]
@mhrimaz 
> The schema that is in https://github.com/aas-core-works/aas-core-codegen/commit/92bcf9ef6301f51f506304f27013f23831238c80 works in python. However, I am not sure if the regex works as intended and covers every corner cases.

No, the schema does not cover the edge cases in Python. The regex pattern operates on UTF-16 strings, but Python encodes characters as UTF-32. Hence, if a string contains no character above 55,639, it should work (no smileys, no Korean characters *etc.*).

For more info, see: https://en.wikipedia.org/wiki/Plane_(Unicode)

@mhrimaz, should I merge then?

**Comment by mristin** [comment]
(
@mhrimaz can you please create another issue on aas-core-codegen for:
> Note that the SHACL has also another critical issue related to DataSpecificationIec61360.
[SHACL states](https://github.com/aas-core-works/aas-core-codegen/blob/92bcf9ef6301f51f506304f27013f23831238c80/test_data/rdf_shacl/test_main/expected/aas_core_meta.v3/expected_output/shacl-schema.ttl#L786) that a rdf:type aas:LangStringDefinitionTypeIec61360 ; should have the following attribute:
<https://admin-shell.io/aas/3/0/LangStringDefinitionTypeIec61360/text>
However in samples it have <https://admin-shell.io/aas/3/0/AbstractLangString/text>. But this is another issue anyways.

?)

**Comment by mhrimaz** [comment]
> Other than this, @mhrimaz, should I merge then?
@mristin Yes, now it is better than before, thanks a lot.

>@mhrimaz can you please create another issue on aas-core-codegen for
I will.

**Comment by mristin** [comment]
@mhrimaz once we fix #519 , we can update the SHACL and RDF schemas in https://github.com/admin-shell-io/aas-specs.
